lib/auth/bot.go

```diff
@@ -0,0 +1,286 @@
+	user, err := types.NewUser(resourceName)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	user.SetRoles([]string{resourceName})
+
+	metadata := user.GetMetadata()
+	metadata.Labels = map[string]string{
+		types.BotLabel:           botName,
+		types.BotGenerationLabel: "0",
+	}
+	user.SetMetadata(metadata)
+
+	// Traits need to be set to silence "failed to find roles or traits" warning
+	user.SetTraits(map[string][]string{
+		teleport.TraitLogins:     {},
+		teleport.TraitKubeUsers:  {},
+		teleport.TraitKubeGroups: {},
+	})
+
+	if err := s.CreateUser(ctx, user); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// createBot creates a new certificate renewal bot from a bot request.
+func (s *Server) createBot(ctx context.Context, req *proto.CreateBotRequest) (*proto.CreateBotResponse, error) {
+	if req.TokenID != "" {
+		// TODO: IAM joining for bots
+		return nil, trace.NotImplemented("IAM join for bots is not yet supported")
+	}
+
+	if req.Name == "" {
+		return nil, trace.BadParameter("bot name must not be empty")
+	}
+
+	resourceName := botResourceName(req.Name)
+
+	// Ensure existing resources don't already exist.
+	_, err := s.GetRole(ctx, resourceName)
+	if err != nil && !trace.IsNotFound(err) {
+		return nil, trace.Wrap(err)
+	}
+	if roleExists := (err == nil); roleExists {
+		return nil, trace.AlreadyExists("cannot add bot: role %q already exists", resourceName)
+	}
+
+	_, err = s.GetUser(resourceName, false)
+	if err != nil && !trace.IsNotFound(err) {
+		return nil, trace.Wrap(err)
+	}
+	if userExists := (err == nil); userExists {
+		return nil, trace.AlreadyExists("cannot add bot: user %q already exists", resourceName)
+	}
+
+	// Create the resources.
+	if err := createBotRole(ctx, s, req.Name, resourceName, req.Roles); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := createBotUser(ctx, s, req.Name, resourceName); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	ttl := time.Duration(req.TTL)
+	if ttl == 0 {
+		ttl = defaults.DefaultBotJoinTTL
+	}
+
+	token, err := s.CreateBotJoinToken(ctx, CreateUserTokenRequest{
+		Name: resourceName,
+		TTL:  ttl,
+		Type: UserTokenTypeBot,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &proto.CreateBotResponse{
+		TokenID:  token.GetName(),
+		UserName: resourceName,
+		RoleName: resourceName,
+		TokenTTL: proto.Duration(ttl),
+	}, nil
+}
+
+// deleteBotUser removes an existing bot user, ensuring that it has bot labels
+// matching the bot before deleting anything.
+func (s *Server) deleteBotUser(ctx context.Context, botName, resourceName string) error {
+	user, err := s.GetUser(resourceName, false)
+	if err != nil {
+		err = trace.WrapWithMessage(err, "could not fetch expected bot user %s", resourceName)
+	} else {
+		label, ok := user.GetMetadata().Labels[types.BotLabel]
+		if !ok {
+			err = trace.Errorf("will not delete user %s that is missing label %s", resourceName, types.BotLabel)
```

Comment:
We don't - I've left a more detailed reply at https://github.com/gravitational/teleport/pull/10099#discussion_r810256017 but while users deleting the labels would cause minor confusion, I think both the error output and resolution steps are reasonably clear.

I'll go ahead and add some more resolution advice to these errors to play it safe, though ("delete the (user|role) manually if desired")